### PR TITLE
[FLINK-8135][REST][docs] Add description to MessageParameter

### PR DESCRIPTION
**Pull request by andrew-half**
## This pull request adds methods for auto-documenting REST API.

Introduced getDescription() method will generate path and query parameter description on 
[REST API documentation](https://ci.apache.org/projects/flink/flink-docs-master/monitoring/rest_api.html) page. 
 
## Brief change log

  - An abstract **getDescription()** method was added to the base **MessageParameter.java** class.
  - All extending classes overrides the method with brief parameter description.
  - The method is called in the **RestAPIDocGenerator.java** class for filling description sections in the output HTML file.

This change is a trivial documentation fix without any test coverage.

## This pull request potentially affect one of the following parts:

  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation
  - Does this pull request introduce a new feature? no
  - This pull request just adds documentation for an existed API

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/AccumulatorsIncludeSerializedValueQueryParameter.java [hunk 1]
```diff
@@ -38,4 +38,10 @@ public String convertValueToString(Boolean value) {
 	public Boolean convertStringToValue(String value) {
 		return Boolean.valueOf(value);
 	}
+
+	@Override
+	public String getDescription() {
+		return "Boolean value. If true then serialized user task accumulators will be included into a response. " +
+			"False by default.";
```
Comment (walterddr) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/AccumulatorsIncludeSerializedValueQueryParameter.java]: "False by default." is confusing to me, is "false": the default value of the parameter is `false`, or "false": whether the task accumulator will be included or not depends on a "default" configuration from the task accumulator?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/SubtaskIndexPathParameter.java [hunk 2]
```diff
@@ -44,4 +44,9 @@ protected String convertToString(final Integer value) {
 		return value.toString();
 	}
 
+	@Override
+	public String getDescription() {
+		return "The index of a subtask. Integer value starts from 0, should be positive.";
```
[the remark this answers is not in the document]
Comment (walterddr) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/SubtaskIndexPathParameter.java]: `must be non-negative`, if "0" is valid

File: flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/MessageParametersTest.java [hunk 3]
```diff
@@ -109,8 +114,13 @@ public JobID convertStringToValue(String value) {
 		}
 
 		@Override
-		public String convertValueToString(JobID value) {
-			return value.toString();
+	public String convertValueToString(JobID value) {
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rest/messages/MessageParametersTest.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/metrics/MetricsAggregationParameter.java [hunk 4]
```diff
@@ -46,6 +46,13 @@ public String convertValueToString(AggregationMode value) {
 		return value.name().toLowerCase();
 	}
 
+	@Override
+	public String getDescription() {
+		return "Comma-separated list of aggregates which should be calculated. Available aggregations are " +
+			" \"sum\", \"max\", \"min\" and \"avg\". Unknown aggregations are ignored. " +
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/metrics/MetricsAggregationParameter.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java [hunk 5]
```diff
@@ -41,6 +41,11 @@ public String convertValueToString(TerminationMode value) {
 		return value.name().toLowerCase();
 	}
 
+	@Override
+	public String getDescription() {
+		return "Defines the termination mode. Supported values are: \"cancel\",\"stop\" (case insensitive). ";
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/SubtaskIndexPathParameter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/SubtaskIndexPathParameter.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/metrics/MetricsAggregationParameter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/metrics/MetricsAggregationParameter.java]: In some cases a user should know that mistakes can be hidden without warnings.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java [hunk 6]
```diff
@@ -30,4 +30,8 @@ public ProgramArgsQueryParameter() {
 		super("program-args", MessageParameterRequisiteness.OPTIONAL);
 	}
 
+	@Override
+	public String getDescription() {
+		return "Specifies the arguments for the program or plan in \"--foo bar\" format.";
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/ProgramArgsQueryParameter.java]: Added examples from JavaDoc. May be you could provide me a batter description?

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerIdPathParameter.java [hunk 7]
```diff
@@ -41,4 +41,9 @@ protected ResourceID convertFromString(String value) {
 	protected String convertToString(ResourceID value) {
 		return value.getResourceIdString();
 	}
+
+	@Override
+	public String getDescription() {
+		return "Task manager ID. Hexadecimal string representation of unique 16 bytes value.";
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/taskmanager/TaskManagerIdPathParameter.java]: Done

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/SavepointPathQueryParameter.java [hunk 8]
```diff
@@ -28,4 +28,9 @@
 	public SavepointPathQueryParameter() {
 		super(KEY, MessageParameterRequisiteness.OPTIONAL);
 	}
+
+	@Override
+	public String getDescription() {
+		return "Defines a path to savepoint to restore from.";
```
[the remark this answers is not in the document]
Comment (andrew-half) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/SavepointPathQueryParameter.java]: Done

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/AccumulatorsIncludeSerializedValueQueryParameter.java (hunk 1 above)
Comment (andrew-half) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/AccumulatorsIncludeSerializedValueQueryParameter.java]: Done

**Comment by andrew-half** [comment]
@walterddr @zentol Thank you guys for a  substantial code-review. Please check the updates then you have time.

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 9]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
 		}
 	}
 
+	/**
+	 * Generates a string containing a comma-separated list of values in double-quatas.
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: typo: quotes

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 10]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
...
+	 * Uses lower-cased values returning from {@link Object#toString()} method for each element in the given array.
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: typo: returned

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 11]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
...
+	 * Null values are spipped.
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: typo: skipped

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 12]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
...
+	 *
+	 * @param values array of elements for the list
+	 *
+	 * @return The string with quated list of elements
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: typo: quoted

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 13]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
...
+	 */
+	public static String toQuatedListString(Object[] values) {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: typo: quoted

Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: We could move this into the `Utils´ class in `flink-docs`, at the moment I don't see much other use for it..

File: flink-core/src/main/java/org/apache/flink/util/StringUtils.java [hunk 14]
```diff
@@ -348,6 +350,21 @@ public static String concatenateWithAnd(@Nullable String s1, @Nullable String s2
...
+		return Arrays.stream(values).filter(Objects::nonNull)
+			.map(v -> v.toString().toLowerCase())
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/util/StringUtils.java]: this reminds me to check whether all enum parameters accept lower-case values...

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/AllowNonRestoredStateQueryParameter.java [hunk 15]
```diff
@@ -42,4 +42,10 @@ public Boolean convertStringToValue(final String value) {
 	public String convertValueToString(final Boolean value) {
 		return value.toString();
 	}
+
+	@Override
+	public String getDescription() {
+		return "Boolean value that specifies whether  non restored state is allowed " +
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/AllowNonRestoredStateQueryParameter.java]: double space after "whether"

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/EntryClassQueryParameter.java [hunk 16]
```diff
@@ -28,4 +28,10 @@
 	public EntryClassQueryParameter() {
 		super("entry-class", MessageParameterRequisiteness.OPTIONAL);
 	}
+
+	@Override
+	public String getDescription() {
+		return "String value specifies the fully qualified name of the entry point class. " +
```
Comment (zentol) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/EntryClassQueryParameter.java]: missing "that" after value, (more occurrences in other files)

**Pull request merged**